File: go/reputation/anomaly_detector.go
// SPDX-License-Identifier: MIT

package reputation

import (
	"sync"
	"time"
)

// AnomalyDetectorOptions configures the thresholds and windows used by
// BehavioralAnomalyDetector.
type AnomalyDetectorOptions struct {
	// VolumeWindowMs is the rolling window length (ms) for volume spike detection.
	VolumeWindowMs int64
	// VolumeSpikeMultiplier is the factor above the EWMA baseline that triggers
	// a flood signal.
	VolumeSpikeMultiplier float64
	// EwmaAlpha is the smoothing factor (0 < α ≤ 1) for the EWMA baseline.
	EwmaAlpha float64
	// ScatterWindowMs is the lookback window (ms) for destination-scatter detection.
	ScatterWindowMs int64
	// ScatterThreshold is the number of unique destinations within ScatterWindowMs
	// that triggers a flood signal.
	ScatterThreshold int
	// GeohashPrefixLength is the number of leading characters compared when
	// checking for a geohash mismatch.
	GeohashPrefixLength int
	// GeohashRateLimitMs is the minimum gap (ms) between successive geohash
	// mismatch signals for the same UHID.  0 means every mismatch fires.
	GeohashRateLimitMs int64
}

// DefaultAnomalyDetectorOptions returns the production defaults.
func DefaultAnomalyDetectorOptions() AnomalyDetectorOptions {
	return AnomalyDetectorOptions{
		VolumeWindowMs:        30_000,
		VolumeSpikeMultiplier: 5.0,
		EwmaAlpha:             0.20,
		ScatterWindowMs:       60_000,
		ScatterThreshold:      50,
		GeohashPrefixLength:   4,
		GeohashRateLimitMs:    60_000,
	}
}

// ---------------------------------------------------------------------------
// per-source internal state
// ---------------------------------------------------------------------------

type volumeState struct {
	mu          sync.Mutex
	windowStart int64 // -1 = not yet initialised
	windowCount int
	ewma        float64
	hasBaseline bool
}

type scatterEntry struct {
	dest string
	ts   int64
}

type scatterState struct {
	mu      sync.Mutex
	entries []scatterEntry
}

type geohashState struct {
	mu           sync.Mutex
	lastSignalMs int64 // -1 = never signalled
}

// ---------------------------------------------------------------------------
// BehavioralAnomalyDetector
// ---------------------------------------------------------------------------

// BehavioralAnomalyDetector observes per-source packet metadata and geohash
// claims, and fires reputation penalties via NodeReputationService when
// anomalies are detected.
//
// All exported methods are safe for concurrent use.
type BehavioralAnomalyDetector struct {
	rep  *NodeReputationService
	opts AnomalyDetectorOptions

	// sync.Map[uhid string -> *volumeState]
	volumeStates sync.Map
	// sync.Map[uhid string -> *scatterState]
	scatterStates sync.Map
	// sync.Map[uhid string -> *geohashState]
	geohashStates sync.Map
}

// NewBehavioralAnomalyDetector returns a new detector using the given
// reputation service and options.
func NewBehavioralAnomalyDetector(rep *NodeReputationService, opts AnomalyDetectorOptions) *BehavioralAnomalyDetector {
	return &BehavioralAnomalyDetector{rep: rep, opts: opts}
}

// ObservePacket records a packet from sourceUhid to destinationUhid at
// timestampMs (Unix milliseconds) and fires reputation penalties if volume-
// spike or destination-scatter anomalies are detected.
func (d *BehavioralAnomalyDetector) ObservePacket(sourceUhid, destinationUhid string, timestampMs int64) {
	d.checkVolume(sourceUhid, timestampMs)
	d.checkScatter(sourceUhid, destinationUhid, timestampMs)
}

// ObserveGeohashClaim compares the first GeohashPrefixLength characters of
// claimedGeohash and observedRoutingGeohash.  If they differ and the rate
// limit allows, RecordSignatureFailure is fired for uhid.
//
// The current wall-clock time (Unix ms) is used as the signal timestamp.
func (d *BehavioralAnomalyDetector) ObserveGeohashClaim(uhid, claimedGeohash, observedRoutingGeohash string) {
	ts := time.Now().UnixMilli()
	d.checkGeohash(uhid, claimedGeohash, observedRoutingGeohash, ts)
}

// ObserveSpkSigFailure directly records a signature failure for uhid.
func (d *BehavioralAnomalyDetector) ObserveSpkSigFailure(uhid string) {
	d.rep.RecordSignatureFailure(uhid)
}

// ---------------------------------------------------------------------------
// volume spike
// ---------------------------------------------------------------------------

func (d *BehavioralAnomalyDetector) getVolumeState(uhid string) *volumeState {
	v, _ := d.volumeStates.LoadOrStore(uhid, &volumeState{windowStart: -1})
	return v.(*volumeState)
}

func (d *BehavioralAnomalyDetector) checkVolume(uhid string, ts int64) {
	vs := d.getVolumeState(uhid)
	vs.mu.Lock()
	defer vs.mu.Unlock()

	// Initialise window on first packet.
	if vs.windowStart == -1 {
		vs.windowStart = ts
		vs.windowCount = 1
		return
	}

	if ts-vs.windowStart >= d.opts.VolumeWindowMs {
		// Roll window: update EWMA baseline.
		if !vs.hasBaseline {
			vs.ewma = float64(vs.windowCount)
			vs.hasBaseline = true
		} else {
			alpha := d.opts.EwmaAlpha
			vs.ewma = alpha*float64(vs.windowCount) + (1-alpha)*vs.ewma
		}

		// Fire flood signal when count exceeded multiplier × positive baseline.
		if vs.hasBaseline && vs.ewma > 0 && float64(vs.windowCount) > d.opts.VolumeSpikeMultiplier*vs.ewma {
			d.rep.RecordRreqFloodAttempt(uhid)
		}

		// Reset window.
		vs.windowStart = ts
		vs.windowCount = 1
	} else {
		vs.windowCount++
	}
}

// ---------------------------------------------------------------------------
// destination scatter
// ---------------------------------------------------------------------------

func (d *BehavioralAnomalyDetector) getScatterState(uhid string) *scatterState {
	v, _ := d.scatterStates.LoadOrStore(uhid, &scatterState{})
	return v.(*scatterState)
}

func (d *BehavioralAnomalyDetector) checkScatter(sourceUhid, destUhid string, ts int64) {
	ss := d.getScatterState(sourceUhid)
	ss.mu.Lock()
	defer ss.mu.Unlock()

	// Append new entry.
	ss.entries = append(ss.entries, scatterEntry{dest: destUhid, ts: ts})

	// Prune entries outside the scatter window.
	cutoff := ts - d.opts.ScatterWindowMs
	trimmed := ss.entries[:0]
	for _, e := range ss.entries {
		if e.ts > cutoff {
			trimmed = append(trimmed, e)
		}
	}
	ss.entries = trimmed

	// Count unique destinations within the window.
	seen := make(map[string]struct{}, len(ss.entries))
	for _, e := range ss.entries {
		seen[e.dest] = struct{}{}
	}
	if len(seen) > d.opts.ScatterThreshold {
		d.rep.RecordRreqFloodAttempt(sourceUhid)
	}
}

// ---------------------------------------------------------------------------
// geohash mismatch
// ---------------------------------------------------------------------------

func (d *BehavioralAnomalyDetector) getGeohashState(uhid string) *geohashState {
	v, _ := d.geohashStates.LoadOrStore(uhid, &geohashState{lastSignalMs: -1})
	return v.(*geohashState)
}

func (d *BehavioralAnomalyDetector) checkGeohash(uhid, claimed, observed string, ts int64) {
	n := d.opts.GeohashPrefixLength
	if geohashPrefix(claimed, n) == geohashPrefix(observed, n) {
		return
	}

	gs := d.getGeohashState(uhid)
	gs.mu.Lock()
	defer gs.mu.Unlock()

	if gs.lastSignalMs == -1 || ts-gs.lastSignalMs >= d.opts.GeohashRateLimitMs {
		d.rep.RecordSignatureFailure(uhid)
		gs.lastSignalMs = ts
	}
}

// ---------------------------------------------------------------------------
// helpers
// ---------------------------------------------------------------------------

// geohashPrefix returns the first n runes of s (or all of s if shorter).
func geohashPrefix(s string, n int) string {
	if n <= 0 {
		return ""
	}
	r := []rune(s)
	if len(r) <= n {
		return s
	}
	return string(r[:n])
}

File: python/aethernet/anomaly_detector.py
# SPDX-License-Identifier: MIT

"""Behavioural anomaly detector — translates mesh traffic patterns into reputation signals.

Detectable patterns:
  1. Volume spike       : packet rate > spike_multiplier × EWMA baseline in a 30 s window
                          → record_rreq_flood_attempt
  2. Destination scatter: single source contacts > scatter_threshold unique destinations
                          in a 60 s sliding window
                          → record_rreq_flood_attempt
  3. Geohash mismatch   : claimed geohash prefix ≠ observed routing prefix
                          → record_signature_failure (rate-limited to 1 signal per 60 s per node)
  4. SPK-sig failure    : direct passthrough
                          → record_signature_failure (no rate limiting)
"""

from __future__ import annotations

import threading
import time
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from aethernet.reputation import NodeReputationService


@dataclass
class AnomalyDetectorOptions:
    """Configurable thresholds for :class:`BehavioralAnomalyDetector`."""

    volume_window_ms: int = 30_000
    """Width of the EWMA volume window in milliseconds."""

    volume_spike_multiplier: float = 5.0
    """Ratio of current window count to EWMA baseline that constitutes a spike."""

    ewma_alpha: float = 0.20
    """Smoothing factor for the exponentially-weighted moving average."""

    scatter_window_ms: int = 60_000
    """Sliding window width (ms) for destination-scatter detection."""

    scatter_threshold: int = 50
    """Number of unique destinations in the window that triggers a flood signal."""

    geohash_prefix_length: int = 4
    """Number of leading characters compared for geohash mismatch detection."""

    geohash_rate_limit_ms: int = 60_000
    """Minimum gap (ms) between geohash mismatch signals for the same node."""


# ---------------------------------------------------------------------------
# Per-node state containers (private)
# ---------------------------------------------------------------------------

@dataclass
class _VolumeState:
    window_start_ms: int | None = None   # None = "no window open yet"
    window_count: int = 0
    ewma_baseline: float = 0.0           # 0.0 means "no completed window yet"


@dataclass
class _ScatterState:
    # List of (destination_uhid, timestamp_ms) — pruned as entries age out
    entries: list[tuple[str, int]] = field(default_factory=list)


# ---------------------------------------------------------------------------
# Main class
# ---------------------------------------------------------------------------

class BehavioralAnomalyDetector:
    """Observes mesh traffic and converts anomalous patterns into reputation signals.

    All public methods are thread-safe.  Per-node state is guarded by individual
    :class:`threading.Lock` objects to minimise contention.

    Args:
        reputation: The :class:`~aethernet.reputation.NodeReputationService` that
            receives the emitted signals.
        options: Optional threshold configuration; uses defaults if *None*.
    """

    def __init__(
        self,
        reputation: "NodeReputationService",
        options: AnomalyDetectorOptions | None = None,
    ) -> None:
        self._reputation = reputation
        self._opts = options or AnomalyDetectorOptions()

        # Per-UHID state — created on first access under _meta_lock
        self._meta_lock = threading.Lock()

        self._volume_states: dict[str, _VolumeState] = {}
        self._volume_locks: dict[str, threading.Lock] = {}

        self._scatter_states: dict[str, _ScatterState] = {}
        self._scatter_locks: dict[str, threading.Lock] = {}

        # Last timestamp at which a geohash mismatch signal was emitted, per UHID
        self._geo_last_signal_ms: dict[str, int] = {}
        self._geo_lock = threading.Lock()

    # ------------------------------------------------------------------ public

    def observe_packet(
        self,
        source_uhid: str,
        destination_uhid: str,
        timestamp_ms: int,
    ) -> None:
        """Record a forwarded/sent packet and check for volume-spike and destination-scatter.

        Args:
            source_uhid:      UHID that originated the packet.
            destination_uhid: UHID that is the packet's next-hop / final destination.
            timestamp_ms:     Wall-clock timestamp when the packet was observed (ms).
        """
        self._check_volume_spike(source_uhid, timestamp_ms)
        self._check_destination_scatter(source_uhid, destination_uhid, timestamp_ms)

    def observe_geohash_claim(
        self,
        uhid: str,
        claimed_geohash: str,
        observed_routing_geohash: str,
    ) -> None:
        """Check whether a node's claimed geohash matches the routing-layer observation.

        Emits :meth:`~aethernet.reputation.NodeReputationService.record_signature_failure`
        when the first ``geohash_prefix_length`` characters differ, subject to a
        per-node rate limit of one signal per ``geohash_rate_limit_ms``.

        Args:
            uhid:                     The node making the claim.
            claimed_geohash:          Geohash string the node advertised.
            observed_routing_geohash: Geohash derived from actual routing behaviour.
        """
        # No caller-supplied timestamp here, so capture the system wall-clock and delegate
        # to the timestamp-aware path. This gives a REAL per-node time window — fire on the
        # first prefix mismatch, suppress repeats within geohash_rate_limit_ms, then reset
        # once the window elapses — instead of the previous fire-once-per-node sentinel that
        # set last_ms=0 and never reset, silently dropping every later mismatch (so a
        # persistent geohash spoofer was penalised exactly once, then ignored forever).
        now_ms = int(time.time() * 1000)
        self.observe_geohash_claim_ts(uhid, claimed_geohash, observed_routing_geohash, now_ms)

    def observe_geohash_claim_ts(
        self,
        uhid: str,
        claimed_geohash: str,
        observed_routing_geohash: str,
        timestamp_ms: int,
    ) -> None:
        """Timestamp-aware variant of :meth:`observe_geohash_claim`.

        Preferred when the caller has a reliable wall-clock value; allows the
        rate-limit window to reset after ``geohash_rate_limit_ms`` has elapsed.

        Args:
            uhid:                     The node making the claim.
            claimed_geohash:          Geohash string the node advertised.
            observed_routing_geohash: Geohash derived from actual routing behaviour.
            timestamp_ms:             Observation time in milliseconds.
        """
        pl = self._opts.geohash_prefix_length
        if claimed_geohash[:pl] == observed_routing_geohash[:pl]:
            return

        with self._geo_lock:
            last_ms = self._geo_last_signal_ms.get(uhid, -1)
            if last_ms == -1 or (timestamp_ms - last_ms) >= self._opts.geohash_rate_limit_ms:
                self._geo_last_signal_ms[uhid] = timestamp_ms
                should_signal = True
            else:
                should_signal = False

        if should_signal:
            self._reputation.record_signature_failure(uhid)

    def observe_spk_sig_failure(self, uhid: str) -> None:
        """Pass a signed-pre-key signature failure directly to the reputation service.

        No rate-limiting is applied; every failure is forwarded immediately.

        Args:
            uhid: The node whose SPK signature failed verification.
        """
        self._reputation.record_signature_failure(uhid)

    # ----------------------------------------------------------------- private

    def _get_or_create_volume_state(self, uhid: str) -> tuple[_VolumeState, threading.Lock]:
        with self._meta_lock:
            if uhid not in self._volume_states:
                self._volume_states[uhid] = _VolumeState()
                self._volume_locks[uhid] = threading.Lock()
            return self._volume_states[uhid], self._volume_locks[uhid]

    def _get_or_create_scatter_state(self, uhid: str) -> tuple[_ScatterState, threading.Lock]:
        with self._meta_lock:
            if uhid not in self._scatter_states:
                self._scatter_states[uhid] = _ScatterState()
                self._scatter_locks[uhid] = threading.Lock()
            return self._scatter_states[uhid], self._scatter_locks[uhid]

    def _check_volume_spike(self, uhid: str, timestamp_ms: int) -> None:
        """EWMA volume-spike detection for *uhid* at *timestamp_ms*."""
        state, lock = self._get_or_create_volume_state(uhid)
        spike_detected = False

        with lock:
            window_ms = self._opts.volume_window_ms
            alpha = self._opts.ewma_alpha
            multiplier = self._opts.volume_spike_multiplier

            if state.window_start_ms is None:
                # First packet ever seen — open the first window
                state.window_start_ms = timestamp_ms
                state.window_count = 1
                return

            if timestamp_ms - state.window_start_ms < window_ms:
                # Still within the current window — increment and check for spike
                state.window_count += 1
                if state.ewma_baseline > 0.0:
                    if state.window_count > multiplier * state.ewma_baseline:
                        spike_detected = True
            else:
                # Window has elapsed — commit it to the EWMA, open a new one
                completed_count = state.window_count
                if state.ewma_baseline == 0.0:
                    # First completed window seeds the EWMA baseline
                    state.ewma_baseline = float(completed_count)
                else:
                    state.ewma_baseline = (
                        alpha * completed_count + (1.0 - alpha) * state.ewma_baseline
                    )

                # Open a new window for this packet
                state.window_start_ms = timestamp_ms
                state.window_count = 1

        if spike_detected:
            self._reputation.record_rreq_flood_attempt(uhid)

    def _check_destination_scatter(
        self,
        source_uhid: str,
        destination_uhid: str,
        timestamp_ms: int,
    ) -> None:
        """Destination-scatter detection for *source_uhid* at *timestamp_ms*."""
        state, lock = self._get_or_create_scatter_state(source_uhid)
        scatter_detected = False

        with lock:
            window_ms = self._opts.scatter_window_ms
            threshold = self._opts.scatter_threshold
            cutoff = timestamp_ms - window_ms

            # Prune entries that have aged out of the sliding window
            state.entries = [
                (dest, ts) for (dest, ts) in state.entries if ts > cutoff
            ]

            # Record this observation
            state.entries.append((destination_uhid, timestamp_ms))

            # Count unique destinations in the window
            unique_dests = {dest for dest, _ in state.entries}
            if len(unique_dests) > threshold:
                scatter_detected = True

        if scatter_detected:
            self._reputation.record_rreq_flood_attempt(source_uhid)
